fontbe/src/features/kern.rs

```diff
@@ -359,31 +390,843 @@ impl Work<Context, AnyWorkId, Error> for KerningGatherWork {
     fn exec(&self, context: &Context) -> Result<(), Error> {
         debug!("Gather be kerning");
         let arc_fragments = context.kern_fragments.all();
+        let ast = context.fea_ast.get();
+        let glyph_order = context.ir.glyph_order.get();
+        let glyph_map = glyph_order.iter().map(|g| g.clone().into_inner()).collect();
         let mut fragments: Vec<_> = arc_fragments
             .iter()
             .map(|(_, fragment)| fragment.as_ref())
             .collect();
         fragments.sort_by_key(|fragment| fragment.segment);
 
-        let mut builder = PairPosBuilder::default();
+        let glyphs_and_gids = glyph_order
+            .iter()
+            .enumerate()
+            .map(|(i, glyphname)| {
+                (
+                    context.ir.glyphs.get(&FeWorkId::Glyph(glyphname.clone())),
+                    GlyphId::new(i as u16),
+                )
+            })
+            .collect::<Vec<_>>();
+        let lookups = self.finalize_kerning(&fragments, &ast.ast, &glyph_map, glyphs_and_gids)?;
+        context.fea_rs_kerns.set(lookups);
+        Ok(())
+    }
+}
+
+impl KerningGatherWork {
+    //. take the kerning fragments and generate the kerning lookups.
+    ///
+    // This includes much of the logic from the ufo2ft KernFeatureWriter
+    fn finalize_kerning(
+        &self,
+        fragments: &[&KernFragment],
+        ast: &ParseTree,
+        glyph_map: &GlyphMap,
+        glyphs: Vec<(Arc<Glyph>, GlyphId)>,
+    ) -> Result<FeaRsKerns, Error> {
+        // ignore diagnostics, they'll get logged during actual GSUB compilation
+        let (compilation, _) = fea_rs::compile::compile::<NopVariationInfo, NopFeatureProvider>(
+            ast,
+            glyph_map,
+            None,
+            None,
+            Opts::new().compile_gpos(false),
+        )
+        .map_err(|err| {
+            Error::FeaCompileError(fea_rs::compile::error::CompilerError::CompilationFail(err))
+        })?;
+
+        let gsub = compilation
+            .gsub
+            .as_ref()
+            .map(write_fonts::dump_table)
+            .transpose()
+            .expect("if this doesn't compile we will already panic when we try to add it to the context");
```

Comment:
Not quite sure I follow. This seems like it should return an Error and if that causes a downstream panic that's a bug?

Reply:
what this comment is saying is that if we panic here (because GSUB fails to pack) it's a case where we are already going to panic, because the way contexts work we unwrap this when we save it to the context.